### PR TITLE
Improve move related error messages

### LINKED ISSUE #46599: MIR borrowck: wrong error for move into closure
```
# ramana at pc in ~/Documents/rust-lang/rust on git:fix-e0504 ✖︎ [18:05:33]
→ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc src/test/compile-fail/E0504.rs -Z borrowck=mir 
error[E0505]: cannot move out of `fancy_num` because it is borrowed
  --> src/test/compile-fail/E0504.rs:23:13
   |
21 |       let fancy_ref = &fancy_num;
   |                       ---------- borrow of `fancy_num` occurs here
22 | 
23 |       let x = move || {
   |  _____________^
24 | |         println!("child function: {}", fancy_num.num); //~ ERROR E0504
25 | |     };
   | |_____^ move out of `fancy_num` occurs here

error: aborting due to previous error


# ramana at pc in ~/Documents/rust-lang/rust on git:fix-e0504 ✖︎ [18:12:36]
→ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc src/test/compile-fail/E0504.rs                 
error[E0504]: cannot move `fancy_num` into closure because it is borrowed
  --> src/test/compile-fail/E0504.rs:24:40
   |
21 |     let fancy_ref = &fancy_num;
   |                      --------- borrow of `fancy_num` occurs here
...
24 |         println!("child function: {}", fancy_num.num); //~ ERROR E0504
   |                                        ^^^^^^^^^ move into closure occurs here

error: aborting due to previous error
```

cc @arielb1 

I am working on this.

### PR DESCRIPTION
**Pull request by vramana**
Fixes #46599

### CODE REVIEW

File: src/test/compile-fail/E0504.rs [hunk 1]
```diff
@@ -8,18 +8,22 @@
...
 // except according to those terms.
 
+// revisions: ast mir
+//[mir]compile-flags: -Z borrowck=mir
+
 struct FancyNum {
     num: u8,
 }
 
 fn main() {
```
Comment (estebank) [inline src/test/compile-fail/E0504.rs]: While you're at this, could we move this file to `test/ui`?

File: src/test/compile-fail/E0504.rs [hunk 2]
```diff
@@ -8,18 +8,22 @@
...
-    let fancy_num = FancyNum { num: 5 };
+    let mut fancy_num = FancyNum { num: 5 };
     let fancy_ref = &fancy_num;
 
     let x = move || {
```
Comment (estebank) [inline src/test/compile-fail/E0504.rs]: Could we add a secondary span label pointing at `move ||` saying "moved into this closure"?

File: src/librustc_mir/borrow_check/mod.rs [hunk 3]
```diff
@@ -849,7 +849,17 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                         }
                         WriteKind::Move => {
                             error_reported = true;
-                            this.report_move_out_while_borrowed(context, place_span, &borrow)
+
+                            // If the move occurs due to a closure, then we want to show the span
+                            // of the place where it's used in the closure.
+                            let closure_span = this.find_closure_span(place_span.1, context.loc);
+                            if let Some((_, var_span)) = closure_span {
```
Comment (estebank) [inline src/librustc_mir/borrow_check/mod.rs]: The first span is the `args_span`, you could use it to add a secondary span label pointing at move || saying "moved into this closure". That way there's no ambiguity (think of the case of nested closures).

File: src/librustc_mir/borrow_check/mod.rs [hunk 4]
```diff
@@ -849,7 +849,17 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
+                                let place_span = (place_span.0, var_span);
+                                this.report_move_into_closure(context, place_span, &borrow);
```
Comment (estebank) [inline src/librustc_mir/borrow_check/mod.rs]: `report_move_into_closure` could be called from `report_move_out_while_borrowed` as it already has the `context` and `place_span`, the only things needed to verify wether it happened to be a closure move.

**Review by estebank — COMMENTED** [review]
Looks good to me! A couple of nitpicks inline. Once addressed, r=me.

**Comment by kennytm** [comment]
Hi @vramana! Would you like to fix the remaining nit picks in https://github.com/rust-lang/rust/pull/47093#pullrequestreview-86206723? Thanks!

**Comment by vramana** [comment]
@kennytm Yes. I was a little busy the past week or so. I will fix the nits soon.

**Comment by pietroalbini** [comment]
Hi @vramana, could you address https://github.com/rust-lang/rust/pull/47093#pullrequestreview-86206723 so the PR can be merged?

**Comment by shepmaster** [comment]
Thank you for your hard work, @vramana ! Unfortunately, we haven't head from you in a few weeks, so we are going to go ahead and close this PR to keep the queue tidy. **Please** feel free to address the feedback and reopen this PR!

**Pull request closed without merging**